Replace `verdi database` commands with `verdi storage`

ramirezfranciscof opened the pull request:
The following specific commands were moved:

 - integrity: the checks that were performed here are no longer relevant.
   There will be a new verdi storage integrity that will perform more
   general checks (once the features of the repository are implemented).

 - migrate: moved to `verdi storage migrate`.

 - summary: information now available through `verdi storage info`.

 - version: information now available through `verdi status`.

The respective tests were adapted as well.

closes #5226

ramirezfranciscof commented:
I think this is ready for review. I moved the stuff around and added a couple of tests to cover some parts of the migration that was not previously being tested (now the only part not being tested is the re-prompting but I'm not sure how to trigger that with the `run_cli_command` fixture). One of the tests (`tests_storage_migrate_raises`) has a slightly awkward mock but I explained there why I had to do it that way.

(@chrisjsewell I put @sphuber as the reviewer because I didn't want to bother you during your holidays, but obviously feel free to take a look too)

sphuber commented on aiida/cmdline/commands/cmd_database.py:
```diff
@@ -8,95 +8,61 @@
 # For further information please visit http://www.aiida.net               #
 ###########################################################################
 """`verdi database` commands."""
+# pylint: disable=unused-argument
```
Wouldn't disable this on the global scope.

ramirezfranciscof replied:
Normally I would agree, but I think it makes sense to make it general here. The reason I need to use this is that all the functions in this module will just be warning without actually doing anything with their arguments (except for `migrate` but since it forwards the call in a way that does not explicitly pass the arguments I am unsure pylint will recognize it 🤔), but I still want the arguments to be there so that users passing them get the deprecation warning and not some wrong argument error. This logic applies to all the functions in the module in general.

Does this make sense or do you still prefer if I put this on each of the commands?

ramirezfranciscof replied:
Bah, unless the deprecation decorator changes this behavior and warns irrespective of what extra parameters I pass?

sphuber replied:
Fair enough, can leave it as is since there is no functional code anymore and we don't expect to add anything in it as it will be removed entirely at some point

sphuber commented on aiida/cmdline/commands/cmd_database.py:
```diff
@@ -8,95 +8,61 @@
-    echo.echo(backend_manager.get_schema_version_database())
+    .. deprecated:: v2.1.0
+    """
+    echo.echo_warning(
```
We have a dedicated decorator for this. Probably best to use that. Its in `aiida.cmdline.utils.decorators`.

ramirezfranciscof replied:
Ok, will do. A couple of questions though:

1.  I see it takes care of the warning, but should I still leave the `.. deprecated:: v2.1.0` in the docstring?
2. Do I just decorate the end commands? Just the group? Both the group and the end commands?
3. I didn't find anything on the docs or the wiki for deprecations of the CLI, should I maybe add this info to the [wiki](https://github.com/aiidateam/aiida-core/wiki/Deprecations-and-API-changes)?

sphuber replied:
1. Yes, I think leaving the explicit deprecated in the docstring is ok. But I think it should mention the version [when it was deprecated](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated), not when it is going to be removed.
2. definitely the leaf commands as the group is not really called (just for the help) but if it works, why not also add it there.
3. Feel free to add it to the wiki. Since this is a developer's aspect, it should go there and not the docs.

ramirezfranciscof replied:
1. I see. But then how do we keep easy track of when we agreed to remove something? I included this info in the deprecation warning now but on second thought I am not sure this is something we want to show to the user necessarily... 🤔 
2. Nah, in the end it did nothing for the groups, not even showed up if I tried using the help. But I think if this is a limitation of the decorator we could maybe extend it to also work on groups no? If it is even possible. For the future anyways.
3. Check out if you think it is clear: [wiki section](https://github.com/aiidateam/aiida-core/wiki/Deprecations-and-API-changes)

sphuber replied:
I think it is fine to mention to the user when we are planning on removing it. Why wouldn't we? This shouldn't be a secret should it and it affects them.

sphuber commented on aiida/cmdline/commands/cmd_database.py:
```diff
@@ -8,95 +8,61 @@
+    """
+    echo.echo_warning(
+        'This command has been deprecated and will be removed soon. '
+        'The same information should now be available in `verdi status`.\n'
```
Wouldn't use the word "should". Either it is or it isn't.

ramirezfranciscof replied:
Done.

sphuber commented on aiida/cmdline/commands/cmd_database.py:
```diff
@@ -119,123 +85,48 @@ def detect_duplicate_uuid(table, apply_patch):
+    .. deprecated:: v3.0.0
+    """
+    echo.echo_warning(
+        'This integrity test has been deemed no longer necessary: it has been deprecated and will be removed soon.\n'
+        'Please check for available tests at `verdi storage integrity` instead.\n'
```
```suggestion
        'This command has been deprecated and will be removed soon.\n'
        'For remaining available integrity checks, use `verdi storage integrity` instead.\n'
```

sphuber commented on aiida/cmdline/commands/cmd_status.py:
```diff
@@ -95,21 +96,31 @@ def verdi_status(print_traceback, no_rmq):
+
     except IncompatibleDatabaseSchema:
-        message = 'Database schema version is incompatible with the code: run `verdi database migrate`.'
+        message = f'Database schema version `Gen {dbgen} - v{dbver}` is incompatible with the code. '
```
The notation `Gen {dbgen} - v{dbver}` will probably be confusing for SqlAlchemy databases. As they are hexadecimal stings, they can start with a letter and it won't be clear that the letter `v` is not part of it but the "version" marker. Maybe just have `Generation/version: {dbgen} / {dbver}`? Think the spaces are important so it is easy to select the version by double clicking

ramirezfranciscof replied:
Done. I didn't use exactly the template you put here but it is virtually the same, let me know what you think.

sphuber replied:
I don't actually see the changes. Did you push it yet?

ramirezfranciscof replied:
Yes, that is weird. You don't see [these](https://github.com/aiidateam/aiida-core/pull/5228/files#diff-98113bf213b93ffbf8b7a306059cea0c28909c38179ef79c148814af298d449dR110-R127)? Github is buggy?

sphuber commented on aiida/cmdline/commands/cmd_storage.py:
```diff
@@ -0,0 +1,125 @@
+
+
+@verdi.group('storage')
+def verdi_database():
```
```suggestion
def verdi_storage():
```

sphuber commented on aiida/cmdline/commands/cmd_storage.py:
```diff
@@ -0,0 +1,125 @@
+@verdi_database.command('migrate')
+@options.FORCE()
+def backend_migrate(force):
+    """Migrate the database to the latest schema version."""
```
```suggestion
    """Migrate the storage to the latest schema version."""
```

sphuber commented on aiida/cmdline/commands/cmd_storage.py:
```diff
@@ -0,0 +1,125 @@
+
+@verdi_database.group('integrity')
+def verdi_database_integrity():
+    """Checks for the integrity of the database and the repository."""
```
```suggestion
    """Checks for the integrity of the data storage."""
```

sphuber commented on aiida/cmdline/commands/cmd_storage.py:
```diff
@@ -0,0 +1,125 @@
+
+@verdi_database.command('migrate')
+@options.FORCE()
+def backend_migrate(force):
```
I think the command functions should just take the name `storage_{command}` where `{command}` is the subcommand.
```suggestion
def storage_migrate(force):
```

sphuber commented on aiida/cmdline/commands/cmd_storage.py:
```diff
@@ -0,0 +1,125 @@
+
+
+@verdi_database.group('integrity')
+def verdi_database_integrity():
```
```suggestion
def storage_integrity():
```

sphuber commented on tests/cmdline/commands/test_storage.py:
```diff
@@ -0,0 +1,118 @@
+    """Test that ``verdi storage migrate`` raises if the daemon is running."""
+    from aiida.engine.daemon.client import DaemonClient
+
+    def return_true():
+        return True
+
+    with monkeypatch.context() as mock_context:
+        mock_context.setattr(DaemonClient, 'is_daemon_running', return_true)
+        result = run_cli_command(cmd_storage.backend_migrate, raises=True)
```
```suggestion
    monkeypatch.setattr(DaemonClient, 'is_daemon_running', lambda: True)
    result = run_cli_command(cmd_storage.backend_migrate, raises=True)
```

ramirezfranciscof replied:
I agree with the `lambda` shortcut, but I think I would still prefer to use the context in general as to be sure I keep the changes "under control". I mean, sure it probably just auto-resets when test function ends and if that is the only procedure in the test it might be not strictly necessary, but I still think is good practice and may help people not familiar with the monkeypatching to understand and extend the test.

ramirezfranciscof replied:
Mmm ok, I have changed if for consistency now, but I am still not convinced this is the best way to do this. I mean, it is true that their doc showcase it without the context and they only introduce the context version for a specific case, but I don't think they explicitly say the other is preferred in general. I just don't see any advantage on the simple statement while I think the context grants increased control in a situation where control is quite useful (ie, when intentionally modifying your classes so that they'll break).

sphuber commented on tests/cmdline/commands/test_storage.py:
```diff
@@ -0,0 +1,118 @@
+        mock_context.setattr(DaemonClient, 'is_daemon_running', return_true)
+        result = run_cli_command(cmd_storage.backend_migrate, raises=True)
+
+    assert 'daemon' in result.output.casefold()
+    assert 'running' in result.output.casefold()
```
if we know that the expected string is ASCII, we don't really need `casefold`. Maybe better to just use lower? Not everyone will be familiar with `casefold` and `lower` is a lot more intuitive what it is doing.

ramirezfranciscof replied:
Done.

sphuber commented on tests/cmdline/commands/test_storage.py:
```diff
@@ -0,0 +1,118 @@
+    """Test that ``verdi storage migrate`` detects the cancelling of the interactive prompt."""
+    import click
+
+    def raise_click_abort(*args, **kwargs):
+        raise click.Abort()
+
+    with monkeypatch.context() as mock_context:
+        mock_context.setattr(click, 'prompt', raise_click_abort)
+        result = run_cli_command(cmd_storage.backend_migrate, raises=True)
```
```suggestion
        monkeypatch.setattr(click, 'prompt', lambda x: raise click.Abort())
        result = run_cli_command(cmd_storage.backend_migrate, raises=True)
```

sphuber replied:
Note I didn't test this, so you may have to massage the lambda definition a bit so it accepts arbitrary args and kwargs but it should work.

ramirezfranciscof replied:
Done. The lambda function apparently [doesn't play along nicely with raises](https://stackoverflow.com/questions/8294618/define-a-lambda-expression-that-raises-an-exception) so I had to use exec, but this is a controlled case where the typical vulnerability of exec does not apply. The line does look a bit clunky, with the need to import the exception and all that, so I left the old way commented out in case we want to use the other one for clarity. Anyways I will have to do at least one more update for the deprecation fixture.

Let me know which you prefer but don't resolve this one yet (hopefully it'll serve as a reminder that I need to remove the comments).

ramirezfranciscof replied:
(removed the commented version now btw)

sphuber commented on tests/cmdline/commands/test_storage.py:
```diff
@@ -0,0 +1,118 @@
+        mock_context.setattr(manager, '_load_backend', return_raiser)
+        result = run_cli_command(cmd_storage.backend_migrate, **call_kwargs)
+
+    assert 'passed error message' in result.output
```
Not sure what you are testing here. You are forcing it to raise an exception with a specific string and then checking that that string is in the output of the CLI command. But that would be a given, wouldn't it? Probably you want to check that the exception is _caught_ and a warning message is shown instead of letting the exception bubble up and show the stack trace. In that case you have to make sure that the output does _not_ contain the stack trace or you have to make your positive check more specific and not just look for just the exception message.

ramirezfranciscof replied:
Well, actually that is a very good question. What is the point of catching [this](https://github.com/aiidateam/aiida-core/pull/5228/files#diff-f3d13a2ae78add210c0cfb8765a16c3de937fd3fe0cac54a35a3d23928e9b953R40-R45)?

``` python
        try:
            backend.migrate()
        except (exceptions.ConfigurationError, exceptions.DatabaseMigrationError) as exception:
            echo.echo_critical(str(exception))
```

It is not adding any contextual information to help the user identify the problem, it is not doing anything afterwards. What is the point of catching to basically re-raise? Just to redirect to the `CMDLINE_LOGGER` in case is not set to stdout/stderr?

sphuber replied:
The point is that these are known failure modes and if they occur, we want to show the error message, but not just have the CLI command except. I catch them explicitly (and not just broad `Exception`) because any other exceptions are not expected and constitute a bug in our code.

sphuber replied:
Yes, perfect.

sphuber commented on tests/cmdline/commands/test_storage.py:
```diff
@@ -0,0 +1,118 @@
+def tests_storage_migrate_raises(run_cli_command, raise_type, call_kwargs, monkeypatch):
+    """Test that ``verdi storage migrate`` detects errors while migrating.
+
+    Note: I can't monkeypatch the result of `manager.get_backend` because the `backend_migrate`
+    function actually calls `manager._load_backend` that will force the re-instantiation of
+    the backend member of the manager, thus overwritting the patch. Instead I monkeypatched
+    the `backend_migrate` directly so that it would return an object whose `migrate` method
+    raises.
```
But it is not really `get_backend` that you want to monkeypatch, no? You want to make `Backend.migrate` raise. So I think you should be able to do something like:
```python
def migrate(self): 
    raise raise_type('passed error message')

monkeypatch.setattr(manage.get_backend().__class__, 'migrate', migrate)
result = run_cli_command(cmd_storage.backend_migrate, **call_kwargs)
```
Need to fetch the class to monkeypatch because the instance will be changed as you mentioned.

ramirezfranciscof replied:
Yes, good point. Done.

sphuber reviewed (changes requested):
Thanks @ramirezfranciscof . Looks good, just some comments.

chrisjsewell commented on aiida/cmdline/commands/cmd_storage.py:
```diff
@@ -0,0 +1,125 @@
+
+@verdi.group('storage')
+def verdi_storage():
+    """Inspect and manage the internal storage."""
```
This feels a bit "non-descript", I don't think a user would really know what this meant.

Perhaps something like

```suggestion
    """Inspect and manage stored data for a profile."""
```

ramirezfranciscof replied:
Done.

ramirezfranciscof commented:
I think I addressed all the comments now (even changing the ones about which I still have some reservations =P). Ready for a second (last? 😬) pass whenever you can @sphuber.

sphuber commented on tests/cmdline/commands/test_status.py:
```diff
@@ -84,3 +86,14 @@ def get_backend():
+    db_ver = backend_manager.get_schema_version_backend()
+
+    result = run_cli_command(cmd_status.verdi_status)
+    assert f'{db_gen} / {db_ver}' in result.output_lines[4]
```
Why not just check `in result.output`? Now you make this test sensitive to added or removed lines, and that is not the point of the test right?

ramirezfranciscof replied:
Mmm no I think the old test was checking the ending of the lines so it was a leftover from that, I agree it is better to make it less sensitive to the rest of the output.

ramirezfranciscof replied:
(which means I changed it)

sphuber commented on aiida/cmdline/commands/cmd_database.py:
```diff
@@ -8,95 +8,60 @@
-            echo.echo_success('migration completed')
+@click.pass_context
+@decorators.deprecated_command(
+    'This command has been deprecated and will be removed soon (in v3.0). '
```
Is it intentional that this will be removed in 3.0 and the rest in 2.1?

ramirezfranciscof replied:
Yes, this is intentional. I personally would prefer to leave the deprecations for longer (not sure if until 3.0, but probably more than 2.1 since I suspect this will be out relatively soon), but this is what was agreed in the last meeting (see [notes](https://docs.google.com/document/d/1WpC-eFK4jXpb581Pfmxv2tLrMMZT9XLg2hH4Z8rQDOo/edit#heading=h.dtdzkrsqjy)).

BTW: I'm still not convinced about exposing this information to the user since this is mostly for internal reference (i.e. remembering when to take it out), but I didn't know where else to track this since the `.. deprecated:: vA.B.C` addition to the docstring is for recording when it was deprecated and not up to when. Any suggestions?

sphuber replied:
I don't agree that it is for internal remembering. Since it affects the user exactly when it is going to be removed, it seems like very relevant information for them. Only thing missing is _when_ the version that it will be removed in is expected to be released.

sphuber reviewed (dismissed):
Thanks @ramirezfranciscof  two minor things. Already approve if it is intentional and don't want to change it, but otherwise I will quickly reapprove after you addressed it.

sphuber reviewed (approved):
Good to go for me, thanks @ramirezfranciscof

ramirezfranciscof commented:
Thanks a lot for the review @sphuber !
